alvr/client_openxr/src/stream.rs

```diff
@@ -290,26 +302,57 @@ impl StreamContext {
         }));
     }
 
+    pub fn initialize_decoder(&mut self, codec: CodecType, config_nal: Vec<u8>) {
+        let config = DecoderConfig {
+            codec,
+            force_software_decoder: self.config.force_software_decoder,
+            max_buffering_frames: self.config.max_buffering_frames,
+            buffering_history_weight: self.config.buffering_history_weight,
+            options: self.config.decoder_options.clone(),
+            config_buffer: config_nal,
+        };
+
+        let (mut sink, source) = decoder::create_decoder(config, {
+            let ctx = Arc::clone(&self.core_context);
+            move |maybe_timestamp: Result<Duration>| match maybe_timestamp {
+                Ok(timestamp) => ctx.report_frame_decoded(timestamp),
+                Err(e) => ctx.report_fatal_decoder_error(&e.to_string()),
+            }
+        });
+        self.decoder_source = Some(source);
+
+        self.core_context
+            .set_decoder_input_callback(Box::new(move |timestamp, buffer| -> bool {
+                sink.push_nal(timestamp, buffer)
+            }));
+    }
+
     pub fn render(
         &mut self,
-        decoded_frame: Option<DecodedFrame>,
+        frame_interval: Duration,
         vsync_time: Duration,
-    ) -> CompositionLayerBuilder {
-        let timestamp;
-        let view_params;
-        let buffer_ptr;
-        if let Some(frame) = decoded_frame {
-            timestamp = frame.timestamp;
-            view_params = frame.view_params;
-            buffer_ptr = frame.buffer_ptr;
-
-            self.last_good_view_params = frame.view_params;
-        } else {
-            timestamp = vsync_time;
-            view_params = self.last_good_view_params;
-            buffer_ptr = std::ptr::null_mut();
+    ) -> (CompositionLayerBuilder, Duration) {
+        let frame_poll_deadline = Instant::now()
+            + Duration::from_secs_f32(
+                frame_interval.as_secs_f32() * DECODER_MAX_TIMEOUT_MULTIPLIER,
+            );
+        let mut frame_result = None;
+        if let Some(source) = self.decoder_source.as_mut() {
+            while frame_result.is_none() && Instant::now() < frame_poll_deadline {
+                frame_result = source.get_frame();
+                thread::yield_now();
```

Comment:
This is actually a busy loop. When there is a healthy amount of buffering, this will succeed immediately without busy looping. The code is the same as before just moved from lib.rs to stream.rs.

Reply:
Something like 200 microsecs or 500 microsecs

Reply:
By reading online, seems that yield() is generally discouraged, mainly because the implementation is platform dependent and even within the same platform is quite unpredictable. The thread could actually resume immediately or wait up to 3ms. So it should never be used unless we did extensive profiling checking it does exactly what we want. It's much better to use our own sleep to balance out CPU usage and latency depending on our needs. So i will change this.